operator/pkg/model/translation/ingress/shared_ingress.go

```diff
@@ -0,0 +1,225 @@
+	portHostNameRouteMap := map[string]map[string][]model.HTTPRoute{}
+	for _, l := range m.HTTP {
+		for _, r := range l.Routes {
+			port := "insecure"
+			if l.TLS != nil {
+				port = "secure"
+			}
+			if _, ok := portHostNameRouteMap[port]; !ok {
+				portHostNameRouteMap[port] = map[string][]model.HTTPRoute{}
+			}
+			portHostNameRouteMap[port][l.Hostname] = append(portHostNameRouteMap[port][l.Hostname], r)
+		}
+	}
+
+	var res []ciliumv2.XDSResource
+	for port, hostNameRouteMap := range portHostNameRouteMap {
+		var virtualhosts []*envoy_config_route_v3.VirtualHost
+
+		// Add HTTPs redirect virtual host for secure host
+		if port == insecureHost && i.enforceHTTPs {
+			for h, r := range portHostNameRouteMap[secureHost] {
+				vhs, _ := translation.NewVirtualHostWithDefaults(h, true, r)
+				virtualhosts = append(virtualhosts, vhs)
+			}
+		}
+
+		for h, r := range hostNameRouteMap {
+			vhs, _ := translation.NewVirtualHostWithDefaults(h, false, r)
+			virtualhosts = append(virtualhosts, vhs)
+		}
+
+		// the route name should match the value in http connection manager
+		// otherwise the request will be dropped by envoy
+		routeName := fmt.Sprintf("%s-%s-listener-%s", i.namespace, i.name, port)
+		rc, _ := translation.NewRouteConfiguration(routeName, virtualhosts)
+		res = append(res, rc)
+	}
+	return res
+}
+
+func (i *SharedIngressTranslator) getClusters(m *model.Model) []ciliumv2.XDSResource {
+	namespaceNamePortMap := getNamespaceNamePortsMap(m)
+
+	var sortedClusterNames []string
+	for ns, v := range namespaceNamePortMap {
+		for name, ports := range v {
+			for _, port := range ports {
+				// the name is having the format of "namespace/name:port"
+				sortedClusterNames = append(sortedClusterNames, fmt.Sprintf("%s/%s:%s", ns, name, port))
+			}
+		}
+	}
+	sort.Strings(sortedClusterNames)
+
+	res := make([]ciliumv2.XDSResource, 0, len(sortedClusterNames))
+	for _, name := range sortedClusterNames {
+		c, _ := translation.NewClusterWithDefaults(name)
+		res = append(res, c)
+	}
+
+	return res
+}
+
+// getNamespaceNamePortsMap returns a map of namespace -> name -> ports.
+// The ports are sorted and unique.
+func getNamespaceNamePortsMap(m *model.Model) map[string]map[string][]string {
+	namespaceNamePortMap := map[string]map[string][]string{}
+	for _, l := range m.HTTP {
+		for _, r := range l.Routes {
+			for _, be := range r.Backends {
+				namePortMap, exist := namespaceNamePortMap[be.Namespace]
+				if exist {
+					namePortMap[be.Name] = sortAndUnique(append(namePortMap[be.Name], be.Port.GetPort()))
+				} else {
+					namePortMap = map[string][]string{
+						be.Name: {be.Port.GetPort()},
+					}
+				}
+				namespaceNamePortMap[be.Namespace] = namePortMap
+			}
+		}
+	}
+	return namespaceNamePortMap
+}
+
+func sortAndUnique(arr []string) []string {
+	m := map[string]struct{}{}
+	for _, s := range arr {
+		m[s] = struct{}{}
+	}
+
+	var res []string
```

Comment:
`res := make([]string, 0, len(m))` should improve allocation efficiency when `m` is not so tiny.